doc/man3/mc_getopt.3
.\"
.\" $Id: mc_getopt.3,v 1.1 2000/07/27 16:59:03 alaffin Exp $
.\"
.\" Copyright (c) 2000 Silicon Graphics, Inc.  All Rights Reserved.
.\" 
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of version 2 of the GNU General Public License as
.\" published by the Free Software Foundation.
.\" 
.\" This program is distributed in the hope that it would be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\" 
.\" Further, this software is distributed without any warranty that it is
.\" free of the rightful claim of any third person regarding infringement
.\" or the like.  Any license provided herein, whether implied or
.\" otherwise, applies only to this software file.  Patent licenses, if
.\" any, provided herein do not apply to combinations of this program with
.\" other software, or any other product whatsoever.
.\" 
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write the Free Software Foundation, Inc., 59
.\" Temple Place - Suite 330, Boston MA 02111-1307, USA.
.\" 
.\" Contact information: Silicon Graphics, Inc., 1600 Amphitheatre Pkwy,
.\" Mountain View, CA  94043, or:
.\" 
.\" http://www.sgi.com 
.\" 
.\" For further information regarding this notice, see: 
.\" 
.\" http://oss.sgi.com/projects/GenInfo/NoticeExplan/
.\"
.TH MC_GETOPT 3 07/25/2000 "Linux Test Project"
.SH NAME
mc_getopt \- multiple character option parsing like getopt(3)
.br
mc_getoptv \- multiple character option parsing like getopt(3)

.SH SYNOPSIS

#include "mc_getopt.h"
.br

char *
.br
\fBmc_getopt\fR(\fIargc, argv, flags, optstring\fR)
.br
int \fIargc\fR;
.br
char \fI*argv[]\fR;
.br
int \fIflags\fR;
.br
char \fI*optstring\fR;

\fBmc_getoptv\fR(\fIargc, argv, flags, nopts, opt_arr\fR)
.br
int \fIargc\fR;
.br
char \fI*argv[]\fR;
.br
int \fIflags\fR;
.br
int \fInopts\fR;
.br
char \fI*opt_arr[]\fR;
.br

extern int \fImc_optind\fR;
.br
extern char *\fImc_optarg\fR;
.br
extern char *\fImc_optopt\fR;

.SH DESCRIPTION
The \fBmc_getopt()\fR and \fBmc_getoptv()\fR functions are command line parsers.
The \fIargc\fR parameter specifies the argument count and the \fIargv\fR parameter
specifies the argument array.  The \fIoptstring\fR argument contains a string
of recognized option strings; if an option string is followed by a
colon, the option takes an argument.  Options are whitespace separated.
A option string can be a single character or a word.

The \fIopt_arr\fR array contains \fInopts\fR elements.  Each option element in
\fIopt_arr\fR must be NULL terminated string; if an option string is followed
by a colon, the option takes an argument.  A option string can be a
single character or a word.  \fIopt_arr\fR must be at least \fInopts\fR elements
in size.

The variable \fImc_optind\fR specifies the index of the next element of the
\fIargv\fR parameter to be processed.  It is initialized to 1, and the
\fBmc_getopt()\fR function updates it as each element of \fIargv\fR is processed.

The \fBmc_getopt()\fR and \fBmc_getoptv()\fR functions returns the \fIoptstring\fR
string that uniquely matches the next element of the \fIargv\fR parameter.
If the option string takes an argument, the \fBmc_getopt()\fR
function sets the \fImc_optarg\fR variable to point to the option string
argument according to the following rules:

.BL
.LI
If the \fIargv\fR's option was the last option string in the string,
\fImc_optarg\fR contains
the next element of the \fIargv\fR array, and the \fImc_optind\fR variable is
incremented by 2.  If the resulting value of \fImc_optind\fR is greater than
or equal to \fIargc\fR, the \fBmc_getopt()\fR function returns an \fBMC_MISSING_OPTARG\fR
.LI
Otherwise, \fImc_optarg\fR points to the string following the option
character in that element of the \fIargv\fR parameter, and the \fImc_optind\fR
variable is incremented by 1.
.LE

If any of the following conditions are true when the \fBmc_getopt()\fR
function is called, the \fBmc_getopt()\fR function returns \fBMC_DONE\fR without
changing the \fImc_optind\fR variable:

.BL
.LI
\fIargv[\fImc_optind\fR] is a null pointer
.LI
*\fIargv[\fImc_optind\fR] is not the character '-'
.LI
\fIargv[\fImc_optind\fR] points to the string "-"
.LE

If the \fIargv[\fImc_optind\fR] parameter points to the "--" string, the \fImc_getopt\fR
function returns \fBMC_DONE\fR after incrementing the \fImc_optind\fR variable.

The \fIflags\fR argument can be used to change how an \fIargv\fR option string
is compared to \fIoptstring\fR.  The \fRflags\fR argument is a bitmask.
The only two bits defined thus far are:

.BL
.LI
\fBMC_FULL_TOKEN_MATCH\fR
.LI
MC_CASE_INSENSITIVE
.LE

How \fIflags\fR bits can be used to affect \fBmc_getopt()'\fRs actions:
.BL
.LI
Without the \fBMC_FULL_TOKEN_MATCH\fR bit set, if the \fIargv\fR string exactly
matches one \fIoptstring\fR option, that option will be returned.
Otherwise, if the \fIargv\fR string matches the beginning of only one
\fIoptstring\fR option, that \fIoptstring\fR option is returned.
.LI
With the \fBMC_FULL_TOKEN_MATCH\fR bit set, the \fIargv\fR string must exactly
match the \fIoptstring\fR, excluding the colon ":".  When word options are
being used, the exact word option string must be in \fIargv\fR to be
recognized as a match.
.LI
Without the MC_CASE_INSENSITIVE bit set, case comparisons are used.
The -I option is different than the -i option.
.LI
With the MC_CASE_INSENSITIVE bit set, all strings are converted to
lower case before comparisons are made.  This means the -I and -i
option are the same.  The returned option string will be all lowercase.
The \fBmc_getoptv()\fR function assumes the option string in the \fIopt_arr\fR
is already converted to lowercase.
.LE

If the \fBmc_getopt(v)\fR function encounters an option string that is not
contained in the \fIoptstring\fR parameter, it returns \fBMC_UNKNOWN_OPTION\fR
string.  If it detects a missing option argument, it returns
\fBMC_MISSING_OPTARG\fR string.  Unlike getopt, \fBmc_getopt()\fR will NOT print
message if an \fIargv\fR option string is not found in \fIoptstring\fR.

If the \fIargv\fR option string matches more than one \fIoptstring\fR strings,
\fBMC_AMBIGUOUS_OPTION\fR string is returned.

\fBmc_getopt\fR and \fBmc_getoptv\fR will not modify \fIargv\fR or \fIoptstring\fR.
\fBmc_getopt\fR will will strdup \fIoptstring\fR and update some character pointers.
The \fIoptstring\fR is parsed only once whenever a \fImc_optind\fR is set to
one or once a \fBMC_DONE\fR was returned.  The strdup space is freed before
a \fBMC_DONE\fR is returned.  This means that \fIoptstring\fR is ignored on all
subsequent calls.  The user is not given a pointer to the strdup'ed space; thus
a user can not free it directly.

.SH EXAMPLES
.nf
.in +3
#include <stdio.h>
#include "mc_getopt.h"

#define ITERATION       "iterations:"
#define FLAG            "flag"

#define OPTSTR  "iterations: flag"

main(argc, argv)
int argc;
char **argv;
{
 char *option;
 int iterate=1;
 int flag=0;
 extern int mc_optind;
 extern char *mc_optarg, *mc_optopt;

 while ( (option=mc_getopt(argc, argv, 0, OPTSTR)) != MC_DONE) {

   if ( strcmp(MC_UNKNOWN_OPTION, option) == 0 ) {
     fprintf(stderr,
       "ERROR - Unknown option: \\"-%s\\"\\n", mc_optopt);
     exit(1);
   } else if ( strcmp(MC_AMBIGIOUS_OPTION, option) == 0 ) {
     fprintf(stderr,
       "ERROR - option \\"-%s\\" is not unique, be more specific\\n",
       mc_optopt);
     exit(1);
   } else if ( strcmp(MC_MISSING_OPTARG, option) == 0 ) {
     fprintf(stderr,
       "ERROR - option \\"-%s\\" is missing its argument\\n",
       mc_optopt);
     exit(1);

   } else if ( strcmp(option, ITERATION) == 0 ) {
     if ( sscanf(mc_optarg, "%i", &iterate) != 1 ) {
       fprintf(stderr,
         "ERROR - option \\"-%s\\" argument must be numeric\\n",
         mc_optopt);
       exit(1);
     }
   } else if ( strcmp(option, FLAG) == 0 ) {
     flag=1;
   }

 } /* end of mc_getopt loop */

 /* ... code ... */

}
.in -3
.fi

.SH "SEE ALSO"
getopt(3), parse_opts(3), 
.\"USC_setup(3).

.SH DIAGNOSTICS
The \fBmc_getopt(v)\fR function returns the next option string specified on
the command line.  The string will be the string as listed in \fIoptstring\fR.

\fBMC_MISSING_OPTARG\fR is returned if the \fBmc_getopt(v)\fR function detects a
missing argument.

\fBMC_UNKNOWN_OPTION\fR is returned if the \fBmc_getopt(v)\fR function encounters an
option string not in the \fIoptstring\fR argument.

\fBMC_AMBIGUOUS_OPTION\fR is returned if the \fBmc_getopt(v)\fR function encounters
option string that matches two or more \fIoptstring\fR argument and there
is not a single exact match.  This condition could occur if
\fIflags\fR argument does not have the \fBMC_FULL_TOKEN_MATCH\fR bit set and
the option string is not a unique string as defined in \fIoptstring\fR.
(i.e.  -it, when optsting contains " iterations:  italic")  It
could also occur if an \fIoptstring\fR option is duplicated
(i.e.  -iterations, when optsting contains " iterations: iterations").

Otherwise, the \fBmc_getopt(v)\fR function returns \fBMC_DONE\fR when all command
line options are parsed.

.SH LIMITATIONS
\fBmc_getopt()\fR has an internal limit of 256 options
allowed in \fIoptstring\fR.

Most users are familiar with \fBgetopt\fR(3) and how it affects
the parsing of command-lines.  In attempting to deal with
word, or multiple character, options, there are a couple
differences between command-lines parsed with \fBgetop\fRt and
\fBmc_getopt()\fR.  Command-lines parsed with \fBmc_getopt()\fR,
options can not be concatenated; each option must be listed beginning 
with "\fI-\fR" minus sign and be whitespace separated.
The option argument can not be concatenated to the option;
there must be a whitespace between the option and its argument.


ChangeLog
LTP-20020307
------------
o	fixed IDcheck.sh to prompt for IDs only when missing (Jay Huie)
o	added time duration option to pan (Jay Huie)
o	fixed bug [ #516577 ] ftruncate03 test (Jay Huie)
o	added 4 more cases to getcwd02 test (Jay Huie)
o	added time-schedule tool by Richard Gooch under sched test
	section (Manoj Iyer)
o	added trace_sched tool under sched test section (Manoj Iyer)
o	fixed bug [ #511427 ] pread02 test (Paul Larson)
o	fixed bug [ #523055 ] sched_getscheduler test (Paul Larson)
o	fixed bug [ #523137 ] sched_setscheduler02 test (Paul Larson)
o	fixed bug [ #525688 ] sendfile01 (Paul Larson)
o	added sigpending02 test to syscalls section (Paul Larson)
o       created a KNOWN-FAILURES file to document known 
        test failures (Robbie Williamson)
o	added sem02 test to ipc section (Robbie Williamson)
o	added SCTP section to network tests (Robbie Williamson)
o	fixed bug [ #505515 ] perf_lan6 test (Robbie Williamson)
o	added disktest by Brent Yardley to io section (Robbie Williamson)
o	closed the following bugs:
	491285, 491286, 506536, 514408 (Paul Larson and Robbie Williamson)


LTP-20020207
------------
o       added support for cross-compiling (Todd Inglett)
o       added LKML's cache_leak testcase to ltp/scratch (Nate Straz)
o       added IPv6 support (Robbie Williamson)
o       added "gethost" to /tools (Robbie Williamson)
o	fixed the race conditions in the float tests and removed 
	the sleeps (Robbie Williamson)
o       enabled non-root make authority (Paul Larson)
o       separated compilation into "make" and "make install" (Paul Larson)
o       added ipc_stress test (Manoj Iyer)
o       added pthreads_stress test (Manoj Iyer)
o       made changes to support architecture independence (Manoj Iyer & Paul Larson)
o	closed the following bugs: 
	504960, 505108, 504613, 504616, 491283, 506689, 508055, 506692, 508074
	491289, 506662, 511383, 511391, 511427, 511494, 504649, 514050, 512953
	(Manoj Iyer, Paul Larson, and Robbie Williamson)	


LTP-20020108
------------
o	fixed IDcheck.sh to work with bash1 (Nate Straz)
o	updated menu script (Robbie Williamson)
o	reset errno to 0 at the beginning of TEST macro (Paul Larson)
o	added floating point stress tests (Bull/IBM)
o	added fsx-linux NFS tests (Apple)	
o	define GLIBC_SIGACTION_BUG for sigaction02 (Robbie Williamson)
o	removed bogus testcase from dup201 test (Paul Larson)
o	fixed recvfrom01 case 5 (Wayne Boyer)
o	workaround for smp issue with waitpid12 (Manfred Spraul)


LTP-20011206
------------
o	necessary users/groups can now be created on installation with 
	user's permission
o	added a simple menu-based interface for running the LTP testsuite
o	fixed negative duration in pan output when -l isn't used
o	new set of tests under fs-maim
o	fixed race condition in nfslock01
o	ar01, ld01, ldd01, nm01, objdump01, and size01 fixed when multiple 
	copies are run simultaneously
o	workaround for SIGTTOU hang in ioctl02
o	shmget03 uses IPC_PRIVATE to make it safe when running multiple copies
o	compiler warnings and other minor errors fixed in many tests

LTP-20011107
------------
o	Many improvements to mtest05 and mtest06 tests
o	new test nfsstress
o	included ver_linux in LTP and made it run at the end of test scripts
o	check for necessary users/groups in Makefile and warn if they don't
	exist
o	documented the users and groups necessary for the testcases to run
o	simplified telnet01 when looking for root prompt
o	removed incorrect testcase from sendfile03
o	fixed modify_ldt01 test problems on newer kernels
o	added setrlimit to unlimit core size in waitpid05 test to work 
	around systems where this is set to 0
o	fixed mmstress pthread hang problem

LTP-20010925
------------
o	Testcases have been organized into categories in a directory 
	structure under the testcases subdirectory
o	About 400 new testcases added covering various system calls
o	Andreas Jaeger got confirmation from the glibc people that the
	ulimit test was broken as a result of some changes made in 
	glibc2.2.5.  Since it's the test that's broken, Nate Straz 
	removed the offending testcase.
o	Updated README and INSTALL files
o	Descriptions of the testcases have been added to the doc 
	directory and are organized by type (syscalls.txt, fs.txt, etc.)

------------
2000-09-18 Juan J. Quintela <quintela@fi.udc.es>
           Aaron Laffin <alaffin@sgi.com>

  * tests/mmap001.c: first checkin of a modified version of
  mmap001.c from the memtest suite.

2000-09-08  Egor Duda <deo@logos-m.ru>

	* tests/fpathconf01.c: Close opened file at exit.
	* tests/select01.c: Ditto.

2000-09-08  Andrea Arcangeli <andrea@suse.de>
  * tests/fcntl09.c, tests/fcntl10.c: Rearrange the F_SETLK test to
  to correctly use F_RDLCK and F_WRLCK.

2000-09-06  Egor Duda  <deo@logos-m.ru>

	* lib/parse_opts.c (parse_opts): Initialize allocated string
	to prevent heap corruption.



lib/mc_getopt.c
/* $Header: /cvsroot/ltp/ltp/lib/Attic/mc_getopt.c,v 1.1 2000/07/27 17:13:18 alaffin Exp $ */

/*
 *	(C) COPYRIGHT CRAY RESEARCH, INC.
 *	UNPUBLISHED PROPRIETARY INFORMATION.
 *	ALL RIGHTS RESERVED.
 */
/**********************************************************
 * 
 *    OS Testing - Cray Research, Inc.
 * 
 *    FUNCTION NAME 	: mc_getopt and mc_getoptv
 * 
 *    FUNCTION TITLE	: multiple character option parsing like getopt(3)
 * 
 *    SYNOPSIS:
 *	char *
 *	mc_getopt(argc, argv, flags, optstring)
 *	int argc;
 *	char *argv[];
 *	int flags;
 *	char *optstring;
 *
 *	mc_getoptv(argc, argv, flags, nopts, opt_arr)
 *	int argc;
 *	char *argv[];
 *	int flags;
 *      int nopts;
 *	char *opt_arr[];
 *
 *	extern int mc_optind;
 *	extern char *mc_optarg;
 *	extern char *mc_optopt;
 *
 * 
 *    AUTHOR		: Richard Logan
 * 
 *    DATE		: 01/96
 * 
 *    DESCRIPTION
 *
 *	The mc_getopt() and mc_getoptv() functions are command line parsers.
 *	The argc parameter specifies the argument count and the argv parameter
 *	specifies the argument array.  The optstring argument contains a string
 *      of recognized option strings; if an option string is followed by a
 *      colon, the option takes an argument.  Options are whitespace seperated.
 *	A option string can be a single character or a word.
 *
 *	The opt_arr array contains nopts elements.  Each option element in
 *      opt_arr must be NULL terminated string; if an option string is followed 
 *      by a colon, the option takes an argument.  A option string can be a 
 *	single character or a word.  opt_arr must be at least nopts elements
 *	in size.
 *
 *	The variable mc_optind specifies the index of the next element of the
 *	argv parameter to be processed.  It is initialized to 1, and the
 *	mc_getopt() function updates it as each element of argv is processed.
 *
 *	The mc_getopt() function returns the optstring string that
 *	uniquely matches the next element of the argv parameter.
 *	If the option string takes an argument, the mc_getopt()
 *      function sets the mc_optarg variable to point to the option string 
 *	argument according to the following rules:
 *
 *	o If the argv's option was the last option string in the string, 
 *	mc_optarg contains
 *	the next element of the argv array, and the mc_optind variable is
 *	incremented by 2.  If the resulting value of mc_optind is greater than
 *	or equal to argc, the mc_getopt() function returns an MC_MISSING_OPTARG
 *
 *	o Otherwise, mc_optarg points to the string following the option
 *	character in that element of the argv parameter, and the mc_optind
 *	variable is incremented by 1.
 *
 *	If any of the following conditions are true when the mc_getopt()
 *	function is called, the mc_getopt() function returns MC_DONE without
 *	changing the mc_optind variable:
 *
 *	o argv[mc_optind] is a null pointer
 *
 *	o *argv[mc_optind] is not the character '-'
 *
 *	o argv[mc_optind] points to the string "-"
 *
 *	If the argv[mc_optind] parameter points to the "--" string, the mc_getopt
 *	function returns MC_DONE after incrementing the mc_optind variable.
 *
 *	The flags argument can be used to change how an argv option string
 *	is compared to optstring.  The flags argument is a bitmask.
 *	The only two bits defined thus far are:
 *	
 *	o MC_FULL_TOKEN_MATCH
 *	o MC_CASE_INSENSITIVE
 *
 *	How flags bits can be used to affect mc_getopt()'s actions:
 *
 *	o Without the MC_FULL_TOKEN_MATCH bit set, if the argv string exactly
 *	matches one optstring option, that option will be returned.  
 *	Otherwise, if the argv string matches the beginning of only one
 *	optstring option, that optstring option is returned.
 *
 *	o With the MC_FULL_TOKEN_MATCH bit set, the argv string must exactly
 *	match the optstring, excluding the colon ":".  When word options are
 *	being used, the exact word option string must be in argv to be
 *	reconized as a match.
 *
 *	o Without the MC_CASE_INSENSITIVE bit set, case comparisons are used.
 *	The -I option is different than the -i option.
 *
 *	o With the MC_CASE_INSENSITIVE bit set, all strings are converted to
 *	lower case before comparisons are made.  This means the -I and -i
 *	option are the same.  The returned option string will be all lowercase.
 *	The mc_getoptv() function assumes the option string in the opt_arr
 *	is already converted to lowercase.
 *
 *	If the mc_getopt(v) function encounters an option string that is not
 *	contained in the optstring parameter, it returns MC_UNKNOWN_OPTION
 *	string.  If it detects a missing option argument, it returns 
 *	MC_MISSING_OPTARG string.  Unlike getopt, mc_getopt() will NOT print
 *	message if an argv option string is not found in optstring.
 *
 *	If the argv option string matches more than one optstring strings,
 *	MC_AMBIGUOUS_OPTION string is returned.
 *
 *      mc_getopt(v) will not modify argv or optstring.  It will strdup optstring
 *	and update some character pointers.  The optstring is parsed only
 *	once whenever a mc_optind is set to one or once a MC_DONE was returned.
 *      The strdup space is freed before a MC_DONE is returned.  This means that
 *	optstring is ignored on all subquent calls.
 *
 *
 *    RETURN VALUES
 *	The mc_getopt(v) function returns the next option string specified on
 *	the command line.  The string will be the string as listed in optstring.
 *
 *	MC_MISSING_OPTARG is returned if the mc_getopt(v) function detects a
 *	missing argument.
 *
 *	MC_UNKNOWN_OPTION is returned if the mc_getopt(v) function encounters an
 *	option string not in the optstring argument.
 *
 *	MC_AMBIGUOUS_OPTION is returned if the mc_getopt(v) function encounters
 *	option string that matches two or more optstring argument and there
 *	is not a single exact match.  This condition could occur if
 *	flags argument does not have the MC_FULL_TOKEN_MATCH bit set and
 *	the option string is not a unique string as defined in optstring.
 *	(i.e.  -it, when optsting contains " iterations:  italic")  It
 *	could also occur if an optstring option is duplicated
 *	(i.e.  -iterations, when optsting contains " iterations: iterations").
 *
 *	Otherwise, the mc_getopt(v) function returns MC_DONE when all command
 *	line options are parsed.
 *
 *    LIMITATIONS
 *	mc_getopt() has an internal limit of 256 options
 *	allowed in optstring.
 *
 *#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#**/

#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <sys/param.h>

#include "mc_getopt.h"

#ifndef DEBUG
#define DEBUG 0		/* used for debugging */
#endif

/*
 * Externally accessable variables.
 */
int mc_optind = 1;	/* index to next argv token */
char *mc_optarg; 	/* pointer to option argument string */

char *mc_optopt;	/* pointer to argv or opt_arg option string */

/*
 * Ensure PATH_MAX is defined.
 */
#ifndef PATH_MAX
#define PATH_MAX	1023
#endif

/*
 * Define the max number of option strings that can exist in optstring.
 */
#define MAX_OPTIONS	256

/*******************************************************************************
 *
 *******************************************************************************/
#if __STDC__
char *
mc_getopt(int argc, char * const argv[], int flags, const char *optstring)
#else
char *
mc_getopt(argc, argv, flags, optstring)
int argc;
char * const argv[];
int flags;
const char *optstring;
#endif  /* __STDC__ */
{

    static char *opt_arr[MAX_OPTIONS]; /* array of pointers to option strings */

    static char *optstr=NULL;  /* strdup'ed optstring, ! NULL, means */
			/* space allocated */
    static int nopts;	/* number of opt_arr elements */

    char *token;	/* used in optstring parsing and string comparision */
    char *cc;		/* used in optstring parsing and converting to lower case */
    int optarr_ind=0;	/* index into opt_arr */
    

#if DEBUG > 0
printf(" Entering mc_getopt, argc = %d, mc_optind = %d, flags = %#o, optstring = '%s'\n",
    argc, mc_optind, flags, optstring);
#endif

    if ( mc_optind >= argc || argv[mc_optind] == NULL ||
	argv[mc_optind][0] != '-' || argv[mc_optind][1] == '\0' ) {

#if DEBUG > 1
printf(" no more options to process\n");
#endif
	if ( optstr != NULL )
	    free(optstr);
	return MC_DONE;

    } else if ( strcmp(argv[mc_optind], "--") == 0 ) {
#if DEBUG > 1
printf(" option --, no more options to process\n");
#endif
	mc_optind++;
	if ( optstr != NULL )
	    free(optstr);
	return MC_DONE;	/* we are done */
    }

    /*
     *  We have new argv - reset info for optstr.
     *  Otherwise old optstr is used.
     */
    if ( optstr == NULL || mc_optind == 1 ) {

        /*
         * copy string so that we don't modify user given string
         */
	if ( optstr != NULL )
	    free(optstr);	/* free old strdup'ed space */

	optstr=strdup(optstring);   /* copy string */
	nopts=0;
	optarr_ind=0;

	token=optstr;
        /*
         * walk through optstr option string (whitespace separated).
         */
        for (;;) {
	
            /* skip preceeding separator chars */
            for (; *token && isspace(*token); token++);
	
	    if ( ! *token )	/* no more tokens */
		break;

            cc = token;	/* start of option string */

            /*
             * find first separator character after token
             */
            for (; *cc && !isspace(*cc); cc++) {
		/*
		 * If doing case insensitive comparisons, make all
		 * optstr characters lowercase.
		 */
		if ( flags & MC_CASE_INSENSITIVE && isupper(*cc) )
		    *cc = tolower(*cc);
	    }

	    opt_arr[optarr_ind++] = token;

	    if ( ! *cc)	/* no more tokens */
		break;
    
            *cc = '\0';   /* null terminate */

	    token = ++cc;

        }  /* end of for loop */

	nopts=optarr_ind;
#if DEBUG > 1
	printf("nopts = %d\n", nopts);

	for(optarr_ind=0; optarr_ind<nopts; optarr_ind++)
   	    printf(" optstring option (index %2d) : %14s\n",
	        optarr_ind, opt_arr[optarr_ind]);
#endif
    }  /* end of processs optstring */

    /*
     * Call mc_getoptv with the optstring in opt_arr array format
     */

    cc=mc_getoptv(argc, argv, flags, nopts, opt_arr);

    if ( optstr != NULL && strcmp(cc, MC_DONE) == 0 )
        free(optstr);

    return cc;
}

/***********************************************************************
 *
 ***********************************************************************/
#if __STDC__
char *
mc_getoptv(int argc, char * const argv[], int flags, int nopts,
					char * const opt_arr[])
#else
char *
mc_getoptv(argc, argv, flags, nopts, opt_arr)
int argc;
char * const argv[];
int flags;
int nopts;
char * const opt_arr[];
#endif  /* __STDC__ */
{
    char *token;        /* used in string comparision */
    char *cc;           /* used in converting to lower case */
    char *cptr;
    int optarr_ind=0;   /* index into opt_arr */

    int found;          /* count of argv option partical matches to optstr options */
    int exact;          /* count of argv option exact matches */
    int cnt;            /* counter */
    int ind;            /* index into opt_arr for matched option */
    int reqarg;		/* if opt_arr[ind] requires an arg */

#if DEBUG > 1
printf(" In mc_getoptv: argv[%d] = %s\n", mc_optind, argv[mc_optind]);
#endif

    if ( mc_optind >= argc || argv[mc_optind] == NULL ||
        argv[mc_optind][0] != '-' || argv[mc_optind][1] == '\0' ) {

#if DEBUG > 1
printf(" no more options to process\n");
#endif
        return MC_DONE;

    } else if ( strcmp(argv[mc_optind], "--") == 0 ) {
#if DEBUG > 1
printf(" option --, no more options to process\n");
#endif
        mc_optind++;
        return MC_DONE; /* we are done */
    }

    if ( flags & MC_CASE_INSENSITIVE ) {
        mc_optopt = &argv[mc_optind][1];

        /*
         * If doing case insensitive comparison,
         * copy argv token and make it all lowercase.
         */
        cc = token = strdup(&argv[mc_optind++][1]);
        for( ;  *cc; cc++)
            if ( isupper(*cc) )
                *cc = tolower(*cc);

    } else {
        token = &argv[mc_optind++][1];
        mc_optopt = token;
    }

    found=0;
    exact=0;
    reqarg=0;

    /*
     * Compare token to each option string.
     * Do not stop once a match is found.  There maybe multiple matches.
     */
    for(optarr_ind=0; optarr_ind<nopts; optarr_ind++) {

	/*
	 * compare opt_arr[optarr_ind] to token
	 * character by character.
	 */

	cc=token;
	cptr=opt_arr[optarr_ind];
	cnt=0;	/* count of match characters */

	while ( *cc && *cptr && *cc++ == *cptr++ ) {
	    cnt++;
	}

        /*
	 * Determine if argv option partially matched opstring option.
	 */
	if ( cnt && *cc == '\0' ) { 
	    /*
	     * Determine if argv option fully matched opstring option.
             * Remember that opt_arr[optarr_ind] can have optional ":".
	     */
	    if ( *cptr == ':' || *cptr == '\0' ) {
                exact++;
                found++;
                ind=optarr_ind;
		if ( *cptr == ':' ) {
		    reqarg=1;
		} else {
		    reqarg=0;
		}

	    /*
	     * If not doing full token matches, count this partial match.
	     */
	    } else if ( ! (flags & MC_FULL_TOKEN_MATCH) ) {
	        found++;
                if ( ! exact ) { /* exact matches are used over partial */
                    ind=optarr_ind;

		    /*
		     * Determine if this option string ends with ":",
		     * requiring an argument.
		     * Find the end of the string first, this
		     * would allow a colon in the string.
		     */
		    while ( *cptr ) 
			cptr++;

		    if ( *(cptr-1) == ':' ) {
			reqarg=1;
		    } else {
			reqarg=0;
		    }
		}
	    }
	}    /* end partial match */
    }   /* end option loop */

#if DEBUG > 2
printf("optarr_ind = %d, ind = %d, found = %d, exact = %d\n",
    optarr_ind, ind, found, exact);
#endif

    /* if doing case insensitive comparison, free space */
    if ( flags & MC_CASE_INSENSITIVE ) {
        free(token);
    }

    if ( found == 0 ) {
#if DEBUG > 1
printf("argv option string not found in optstr returning MC_UNKNOWN_OPTION :%s\n", 
MC_UNKNOWN_OPTION);
#endif
        /* argv option string not found in optstr */
        return MC_UNKNOWN_OPTION;

    } else if ( exact == 1 || found == 1 ) {
        /* option is uniquely matched */
#if DEBUG > 1
printf("argv option found a string match\n");
#endif


        if ( reqarg ) { 	/* option requires argument */

            /* If option argument is missing, return MC_MISSING_OPTARG */
            if ( mc_optind >= argc ) {
#if DEBUG > 2
printf("returning MC_MISSING_OPTARG: %s\n", MC_MISSING_OPTARG);
#endif
                return MC_MISSING_OPTARG;
            }

            mc_optarg = argv[mc_optind++];

        }

#if DEBUG > 2
printf("returning the option (index %d) %s\n", ind, opt_arr[ind]);
#endif
        /* return the full mc_getopt option string */
        return opt_arr[ind];

    } else {

        /*
         * This condition will be encountered if
         * argv option string matches more than optstr option string.
         * The user could not have entered a unique option string or
         * the optstr has a duplicate option entered.
         */
#if DEBUG > 1
printf("argv option string not unique\n");
#endif
        return MC_AMBIGUOUS_OPTION;
    }
}



#ifdef UNIT_TEST

/****
 ****  The following code is provided as a basic unit test.
 ****  Compile this code with UNIT_TEST set and a executable
 ****  command will be generated.  The optstring is hardcoded,
 ****  but the programs argc and argv are given to mc_getopt.
 ****/

main(argc, argv)
int argc;
char **argv;
{

#define	OPTSTR1		" iterate: Testpause I: i "
    extern int mc_optind;
    extern char *mc_optarg;

    char *ret;
    int ind=0;

	mc_optind=1;
	mc_optarg=NULL;
	printf("*********** flags = 0, optstr = %s\n", OPTSTR1);
	while ( (ret=mc_getopt(argc, argv, 0, OPTSTR1)) != MC_DONE ) {
	   if ( mc_optarg != NULL )
	        printf("mc_getopt returned %16s, mc_optopt = %8s, mc_optind = %d, mc_optarg = %s\n",
		    ret, mc_optopt, mc_optind, mc_optarg);
	   else
	        printf("mc_getopt returned %16s, mc_optopt = %8s, mc_optind = %d, mc_optarg = NULL\n",
		    ret, mc_optopt, mc_optind);
	   ind++;
	   mc_optarg=NULL;
	}
	if ( mc_optind < argc )
	    printf("mc_getopt is done,  mc_optind = %d (argv[%d] = %s)\n",
		mc_optind, mc_optind, argv[mc_optind]);
	else
	    printf("mc_getopt is done,  mc_optind = %d (no arguments)\n", mc_optind);

	printf("******************************************************\n");

	
	mc_optind=1;
	mc_optarg=NULL;
	printf("*********** flags = MC_FULL_TOKEN_MATCH, optstr = %s\n", OPTSTR1);
	while ( (ret=mc_getopt(argc, argv, MC_FULL_TOKEN_MATCH, OPTSTR1)) != MC_DONE ) {
	   if ( mc_optarg != NULL )
	        printf("mc_getopt returned %16s, mc_optopt = %8s, mc_optind = %d, mc_optarg = %s\n",
		    ret, mc_optopt, mc_optind, mc_optarg);
	   else
	        printf("mc_getopt returned %16s, mc_optopt = %8s, mc_optind = %d, mc_optarg = NULL\n",
		    ret, mc_optopt, mc_optind);
	   mc_optarg=NULL;
	}
	if ( mc_optind < argc )
	    printf("mc_getopt is done,  mc_optind = %d (argv[%d] = %s)\n",
		mc_optind, mc_optind, argv[mc_optind]);
	else
	    printf("mc_getopt is done,  mc_optind = %d (no arguments)\n", mc_optind);

	mc_optind=1;
	mc_optarg=NULL;
	printf("*********** flags = MC_CASE_INSENSITIVE, optstr = %s\n", OPTSTR1);
	while ( (ret=mc_getopt(argc, argv, MC_CASE_INSENSITIVE, OPTSTR1)) != MC_DONE ) {
	   if ( mc_optarg != NULL )
	        printf("mc_getopt returned %16s, mc_optopt = %8s, mc_optind = %d, mc_optarg = %s\n",
		    ret, mc_optopt, mc_optind, mc_optarg);
	   else
	        printf("mc_getopt returned %16s, mc_optopt = %8s, mc_optind = %d, mc_optarg = NULL\n",
		    ret, mc_optopt, mc_optind);
	   mc_optarg=NULL;
	}
	if ( mc_optind < argc )
	    printf("mc_getopt is done,  mc_optind = %d (argv[%d] = %s)\n",
		mc_optind, mc_optind, argv[mc_optind]);
	else
	    printf("mc_getopt is done,  mc_optind = %d (no arguments)\n", mc_optind);

	return 0;
}

/**
Use the following unit test cmd line arguments:
 -i -itera -4 -I -5 -i -IteRate -6 arg

 ***/

#endif /* UNIT_TEST */
